README.md
TestChecks
==============
version: `0.3.0`

Utilities for running JUnit tests against collections of inputs and expected outputs.
Designed to help decouple test data from test logic.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project does its best to adhere to [Semantic Versioning](http://semver.org/).


--------
### [0.3.1](N/A) - 2019-03-25
#### Changed
* Renamed `ObjectMethodTests` -> `ObjectEquality` and setup public static testEquality() method to test `equals()` and `hashCode()` object contracts


--------
### [0.3.0](https://github.com/TeamworkGuy2/TestChecks/commit/12fa4dd54bc0dbbdc8530ca4d1eb02683218f7bc) - 2018-03-25
#### Added
* `TestData.fromInputsAndExpected(Iterable, Iterable, BiFunction)` and overloads for converting sequences of test values into TestData objects

#### Changed
* Renamed package `twg2.testAssist.checks` -> `twg2.junitassist.checks`
* CheckTask `assertTests(Collection, Function)` -> `assertTests(Iterable, Function)`
* CheckCollections `assertLooseEqualsDepth2(Collection, Collection)` -> `assertLooseEqualsNested(Iterable, Iterable)`

#### Removed
* `CheckCollections.toList()`


--------
### [0.2.0](N/A) - 2016-09-18
#### Changed
* Renamed TestDataObj -> TestDataImpl
* Moved some of the Check methods into a new CheckArrays class

#### Fixed
* A array comparison bug in CheckTask.assertTests()


--------
### [0.1.0](N/A) - 2016-06-23
#### Added
* Initial version tracking, includes Check, CheckCollections, CheckTask, ObjectMethodTests, TestData, and TestDataObj


src/twg2/junitassist/checks/Check.java
package twg2.junitassist.checks;

import java.util.Arrays;

import org.junit.Assert;

/** Static methods for checking type, bounds, and other requirements
 * @author TeamworkGuy2
 * @since 2014-6-3
 */
public final class Check {

	/** Throw an array index out of bounds exception if the {@code index} is less
	 * than {@code 0} or greater than or equal to {@code length} 
	 * @param index the index to check
	 * @param length the maximum value of the index (exclusive)
	 * @throws ArrayIndexOutOfBoundsException if the index is outside the range specified
	 */
	public static final void bounds(int index, int length) {
		if(index < 0 || index > length) {
			throw new ArrayIndexOutOfBoundsException(index + ", length=" + length);
		}
	}


	/** Throw an index out of bounds exception if the {@code index} is less
	 * than {@code min} or greater than or equal to {@code max} 
	 * @param index the index to check
	 * @param min the minimum value of the index (inclusive)
	 * @param max the maximum value of the index (exclusive)
	 * @throws IndexOutOfBoundsException if the index is outside the range specified
	 */
	public static final void bounds(int index, int min, int max) {
		if(index < min || index >= max) {
			throw new IndexOutOfBoundsException(index + ", min" + min + ", max=" + max);
		}
	}


	/** Clamp the specified value to the specified range. For example:<br/>
	 * {@code clamp(24, 0, 10) returns 10}<br/>
	 * {@code clamp(4, 1, 20) returns 4}<br/>
	 * {@code clamp(-31, 0, 100) returns 0}
	 * @param num the number to clamp
	 * @param min the minimum value
	 * @param max the maximum value
	 * @return {@code num}, or {@code min} if {@code num} is less than {@code min},
	 * or {@code max} if {@code num} is greater than {@code max}
	 */
	public static final int clamp(int num, int min, int max) {
		return Math.max(Math.min(num, max), min);
	}


	/**
	 * @see #clamp(int, int, int)
	 */
	public static final float clamp(float num, float min, float max) {
		return Math.max(Math.min(num, max), min);
	}


	/**
	 * @see #clamp(int, int, int)
	 */
	public static final long clamp(long num, long min, long max) {
		return Math.max(Math.min(num, max), min);
	}


	/**
	 * @see #clamp(int, int, int)
	 */
	public static final double clamp(double num, double min, double max) {
		return Math.max(Math.min(num, max), min);
	}


	/** Throw a null pointer exception if the object is null
	 * @param obj the object to check
	 * @throws NullPointerException if the object is null
	 */
	public static final void nullArg(Object obj) {
		if(obj == null) {
			throw new NullPointerException();
		}
	}


	/** Check if the string is null, length zero, or contains only whitespace characters.<br/>
	 * Equivalent to: {@code str == null ? true : (str.trim().length() == 0 ? true : false);}
	 * @param str the string to check
	 * @return true if the string is null, length zero, or contains only whitespace characters, false otherwise
	 */
	public static final boolean isEmpty(String str) {
		if(str == null || str.length() == 0) { return true; }
		int len = str.length();
		int start = 0;
		
		while((start < len) && (str.charAt(start) <= ' ')) {
			start++;
		}
		while((start < len) && (str.charAt(len - 1) <= ' ' )) {
			len--;
		}
		if(start == len) { return true; }
		return false;
	}


	/** Return an empty string if the string is null, length zero, or contains only whitespace characters.<br/>
	 * Equivalent to: {@code str == null ? "" : (str.trim().length() == 0 ? "" : str);}
	 * @param str the string to check
	 * @return an empty string ({@code ""}) if the string is null, length zero, or contains only
	 * whitespace characters, returns {@code str} otherwise
	 */
	public static final String checkEmpty(String str) {
		if(str == null || str.length() == 0) { return ""; }
		int len = str.length();
		int start = 0;
		
		while((start < len) && (str.charAt(start) <= ' ')) {
			start++;
		}
		while((start < len) && (str.charAt(len - 1) <= ' ' )) {
			len--;
		}
		if(start == len) { return ""; }
		return str;
	}


	/** Check if two strings are equal
	 * @param str the first string to compare
	 * @param str2 the second string to compare
	 * @param msg a message to include in the error if the two strings are not equal
	 * @throws Error if {@code str} does not equal {@code str2}
	 */
	public static final void checkEqual(String str, String str2, String msg) {
		if(!str.equals(str2)) {
			throw new Error("input: '" + str + "' does not match expected: '" + str2 + "', " + msg);
		}
	}


	public static final void assertEqual(String str, String str2, String msg) {
		Assert.assertEquals("input: '" + str + "' does not match expected: '" + str2 + "', " + msg, str, str2);
	}


	@SafeVarargs
	public static final void assertLengths(int... lens) {
		assertLengths((String)null, lens);
	}


	@SafeVarargs
	public static final void assertLengths(String errMsg, int... lens) {
		if(lens == null || lens.length > 1) {
			return;
		}
		int firstLen = lens[0];
		for(int i = 1, size = lens.length; i < size; i++) {
			int lenI = lens[i];
			if(lenI != firstLen) {
				Assert.assertTrue(errMsg != null ? errMsg : "array " + i + " length (" + lenI + ") not equal expected length (" + firstLen + ")", false);
			}
		}
	}


	// package-private
	static final void expectLength(int len1, int len2) {
		if(len1 != len2) {
			throw new IllegalArgumentException("inputs (length: " + len1 + ") should be equal in length to (length: " + len2 + ")");
		}

	}


	// package-private
	static final void assertLengthEqual(int len1, int len2) {
		Assert.assertEquals("inputs (length: " + len1 + ") should be equal in length to (length: " + len2 + ")", len1, len2);
	}


	// package-private
	static final <T> String toDescriptiveString(T obj) {
		if(obj.getClass().isArray()) {
			if(obj.getClass().getComponentType() == Boolean.TYPE) { return Arrays.toString((boolean[])obj); }
			else if(obj.getClass().getComponentType() == Character.TYPE) { return Arrays.toString((char[])obj); }
			else if(obj.getClass().getComponentType() == Byte.TYPE) { return Arrays.toString((byte[])obj); }
			else if(obj.getClass().getComponentType() == Short.TYPE) { return Arrays.toString((short[])obj); }
			else if(obj.getClass().getComponentType() == Integer.TYPE) { return Arrays.toString((int[])obj); }
			else if(obj.getClass().getComponentType() == Long.TYPE) { return Arrays.toString((long[])obj); }
			else if(obj.getClass().getComponentType() == Float.TYPE) { return Arrays.toString((float[])obj); }
			else if(obj.getClass().getComponentType() == Double.TYPE) { return Arrays.toString((double[])obj); }
			return Arrays.toString((Object[])obj);
		}
		return obj.toString();
	}

}


src/twg2/junitassist/checks/TestData.java
package twg2.junitassist.checks;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Spliterators;
import java.util.function.BiFunction;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

/**
 * @author TeamworkGuy2
 * @since 2015-6-26
 * @param <I> the input data type
 * @param <R> the result data type
 */
public interface TestData<I, R> {

	public I getInput();

	public void setInput(I input);

	public R getExpected();

	public void setExpected(R expected);

	public R getResult();

	public void setResult(R result);

	public boolean isShouldInputEqualExpected();

	public void setShouldInputEqualExpected(boolean shouldInputEqualExpected);

	public boolean isExpectThrow();

	public void setExpectThrow(boolean expectThrow);


	public static <I, R> TestData<I, R> matchTrue(I input, R expected) {
		return new TestDataImpl<>(input, expected, null, true, false);
	}


	public static <I, R> TestData<I, R> matchFalse(I input, R expected) {
		return new TestDataImpl<>(input, expected, null, false, false);
	}


	public static <I, R> TestData<I, R> shouldThrow(I input, R expected) {
		return new TestDataImpl<>(input, expected, null, false, true);
	}


	/** Convert a sequence of input and expected data entries to {@link TestData} objects
	 * @param inputsAndExpected sequence of input and expected data entries
	 * @param createTestData for example {@link TestData#matchTrue(Object, Object)}
	 * @return
	 */
	public static <I, R> List<TestData<I, R>> fromInputsAndExpected(Iterable<? extends Map.Entry<? extends I, ? extends R>> inputsAndExpected, BiFunction<I, R, ? extends TestData<I, R>> createTestData) {
		Stream<Map.Entry<? extends I, ? extends R>> iter = StreamSupport.stream(Spliterators.spliteratorUnknownSize(inputsAndExpected.iterator(), 0), false);

		return fromInputsAndExpected(
				iter.map((e) -> e.getKey()).iterator(),
				iter.map((e) -> e.getValue()).iterator(),
				createTestData);
	}


	/** Convert a sequence of inputs and a sequence of expected data to {@link TestData} objects
	 * @param inputs sequence of input data elements
	 * @param expected sequence of expected data elements
	 * @param createTestData for example {@link TestData#matchTrue(Object, Object)}
	 * @return
	 */
	public static <I, R> List<TestData<I, R>> fromInputsAndExpected(Iterable<? extends I> inputs, Iterable<? extends R> expected, BiFunction<I, R, ? extends TestData<I, R>> createTestData) {
		return fromInputsAndExpected(inputs.iterator(), expected.iterator(), createTestData);
	}


	/** Convert a sequence of inputs and a sequence of expected data to {@link TestData} objects
	 * @param inputs sequence of input data elements
	 * @param expected sequence of expected data elements
	 * @param createTestData for example {@link TestData#matchTrue(Object, Object)}
	 * @return
	 */
	public static <I, R> List<TestData<I, R>> fromInputsAndExpected(Iterator<? extends I> inputsIter, Iterator<? extends R> expectedIter, BiFunction<I, R, ? extends TestData<I, R>> createTestData) {
		List<TestData<I, R>> res = new ArrayList<>();
		while(inputsIter.hasNext() && expectedIter.hasNext()) {
			I input = inputsIter.next();
			R expect = expectedIter.next();
			res.add(createTestData.apply(input, expect));
		}
		if(inputsIter.hasNext()) {
			throw new IllegalArgumentException("'expected' sequence shorter than 'inputs' sequence");
		}
		if(expectedIter.hasNext()) {
			throw new IllegalArgumentException("'inputs' sequence shorter than 'expected' sequence");
		}
		return res;
	}

}


src/twg2/junitassist/checks/CheckTask.java
package twg2.junitassist.checks;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.BiFunction;
import java.util.function.Function;

import org.junit.Assert;

/**
 * @author TeamworkGuy2
 * @since 2015-5-26
 */
public final class CheckTask {
	private static float FLOAT_EPSILON = 0;
	private static double DOUBLE_EPSILON = 0;


	/** Assert that the task throws an {@link Exception}
	 * @param task the action to perform
	 */
	public static final void assertException(Runnable task) {
		assertException(null, task);
	}


	/** Assert that the task throws an {@link Exception}
	 * @param errMsg a message to include with the assertion
	 * @param task the action to perform
	 */
	public static final void assertException(String errMsg, Runnable task) {
		Throwable error = null;
		try {
			task.run();
		} catch(Exception e) {
			error = e;
		} finally {
			Assert.assertTrue("task " + (errMsg != null ? "'" + errMsg + "' " : "") + "was expected to throw exception but did not", error != null);
		}
	}


	/** Assert that the task throws a {@link Throwable}
	 * @param task the action to perform
	 */
	public static final void assertThrowable(Runnable task) {
		assertThrowable(null, task);
	}


	/** Assert that the task throws a {@link Throwable}
	 * @param errMsg a message to include with the assertion
	 * @param task the action to perform
	 */
	public static final void assertThrowable(String errMsg, Runnable task) {
		Throwable error = null;
		try {
			task.run();
		} catch(Throwable e) {
			error = e;
		} finally {
			Assert.assertTrue("task " + (errMsg != null ? "'" + errMsg + "' " : "") + "was expected to throw error but did not", error != null);
		}
	}


	// ==== list wrappers ====
	public static final <T, R> void assertTests(Iterable<T> inputs, Iterable<R> expected,
			BiFunction<T, R, String> errorMessageGen, Function<T, R> action) {
		assertTests(toArray(inputs), toArray(expected), errorMessageGen, action);
	}


	public static final <T, R> void assertTests(Iterable<T> inputs, Iterable<R> expected,
			BiFunction<T, R, String> errorMessageGen, BiFunction<T, Integer, R> action) {
		assertTests(toArray(inputs), toArray(expected), errorMessageGen, action);
	}


	public static final <T, R> void assertTests(Iterable<T> inputs, Iterable<R> expected, Function<T, R> action) {
		assertTests(toArray(inputs), toArray(expected), null, action);
	}


	public static final <T, R> void assertTests(Iterable<T> inputs, Iterable<R> expected, BiFunction<T, Integer, R> action) {
		assertTests(toArray(inputs), toArray(expected), null, action);
	}


	// ==== defaults ====
	public static final <T, R> void assertTests(T[] inputs, R[] expected, Function<T, R> action) {
		assertTests(inputs, expected, null, action);
	}


	public static final <T, R> void assertTests(T[] inputs, R[] expected, BiFunction<T, Integer, R> action) {
		assertTests(inputs, expected, null, action);
	}


	/** Run a series of test against a set of inputs and expected results.
	 * Each input is converted to a result via a function and the result is compared
	 * to the expected result using {@link #equals(Object)}
	 * @param inputs the array of inputs to process
	 * @param expected the expected results of processing each input
	 * @param errorMessageGen the message generate to run if the result of one of
	 * the inputs does not match the expected result 
	 * @param action the action that takes an input from {@code inputs} and converts
	 * it to a result value to compare to {@code expected}
	 */
	public static final <T, R> void assertTests(T[] inputs, R[] expected, BiFunction<T, R, String> errorMessageGen, Function<T, R> action) {
		assertTests(inputs, expected, errorMessageGen, (input, idx) -> action.apply(input));
	}


	public static final <T, R> void assertTests(T[] inputs, R[] expected, BiFunction<T, R, String> errorMessageGen,
			BiFunction<T, Integer, R> action) {
		Check.assertLengthEqual(inputs.length, expected.length);

		for(int i = 0, size = inputs.length; i < size; i++) {
			T input = inputs[i];
			R expect = expected[i];
			R res = action.apply(input, i);

			String errorMsg = (errorMessageGen != null ? errorMessageGen.apply(input, expect) : "") + ", input " + i + ": '" + Check.toDescriptiveString(res) +
					"' does not match expected: '" + Check.toDescriptiveString(expect) + "'";
			// if the results are arrays, compare arrays, else compare objects
			if(CheckArrays.isArrays(expect, res)) {
				CheckArrays.assertEquals(errorMsg, expect, res, FLOAT_EPSILON, DOUBLE_EPSILON);
			}
			else {
				Assert.assertEquals(errorMsg, expected[i], res);
			}
		}
	}


	public static final <I, E, R> void assertTests(Iterable<? extends TestData<I, R>> testData, Function<I, R> action) {
		@SuppressWarnings("rawtypes")
		List<TestData<I, R>> data = (testData instanceof List ? new ArrayList<>(((List)testData).size()) : new ArrayList<>());
		for(TestData<I, R> dat : testData) {
			data.add(dat);
		}
		@SuppressWarnings("unchecked")
		TestData<I, R>[] dataAry = data.toArray(new TestData[data.size()]);
		assertTests(dataAry, (BiFunction<I, R, String>)null, (input, idx) -> action.apply(input));
	}


	public static final <I, E, R> void assertTests(TestData<I, R>[] testData,
			BiFunction<I, R, String> errorMessageGen, Function<I, R> action) {
		assertTests(testData, errorMessageGen, (input, idx) -> action.apply(input));
	}


	public static final <I, R> void assertTests(TestData<I, R>[] testData,
			BiFunction<I, R, String> errorMessageGen, BiFunction<I, Integer, R> action) {
		for(int i = 0, size = testData.length; i < size; i++) {
			final int counter = i;
			final TestData<I, R> dat = testData[i];
			final AtomicReference<R> resRef = new AtomicReference<>();

			if(dat.isExpectThrow()) {
				CheckTask.assertException(() -> resRef.set(action.apply(dat.getInput(), counter)));
			}
			else {
				resRef.set(action.apply(dat.getInput(), i));
				R res = resRef.get();
				R expected = dat.getExpected();
				String msg = (errorMessageGen != null ? errorMessageGen.apply(dat.getInput(), dat.getExpected()) : "") + ", input " + i + ": '" + Check.toDescriptiveString(res) +
						"' does not match expected: '" + Check.toDescriptiveString(dat.getExpected()) + "'";

				if(dat.isShouldInputEqualExpected()) {
					if(CheckArrays.isArrays(expected, res)) {
						CheckArrays.assertEquals(msg, expected, res, FLOAT_EPSILON, DOUBLE_EPSILON);
					}
					else {
						Assert.assertEquals(msg, expected, res);
					}
				}
				else {
					// TODO JUnit doesn't have assertArrayNotEquals() methods, so we'll just assume if the objects are different they don't have the same content if they're arrays
					Assert.assertNotEquals(msg, expected, res);
				}
			}
		}
	}


	private static final <T> T[] toArray(Iterable<T> iter) {
		List<T> values = new ArrayList<>();
		for(T value : iter) {
			values.add(value);
		}
		@SuppressWarnings("unchecked")
		T[] res = (T[])values.toArray();
		return res;
	}

}


src/twg2/junitassist/checks/TestDataImpl.java
package twg2.junitassist.checks;

/** Object for storing test case input data, result data, expected data, and flags indicating what test result is expected
 * @param <I> the input data type
 * @param <R> the comparison result data type
 * @author TeamworkGuy2
 * @since 2014-12-19
 */
public class TestDataImpl<I, R> implements TestData<I, R> {
	private I input;
	private R expected;
	private R result;
	private boolean shouldInputEqualExpected;
	private boolean expectThrow;


	/** Create a test object with input and result data
	 * @param input the input data used by the test
	 * @param expected the expected result
	 * @param result the actual result
	 * @param shouldInputEqualExpected whether the input should equal the expected result
	 * @param expectThrow whether the test is expected to throw an exception
	 */
	public TestDataImpl(I input, R expected, R result, boolean shouldInputEqualExpected, boolean expectThrow) {
		this.input = input;
		this.expected = expected;
		this.result = result;
		this.shouldInputEqualExpected = shouldInputEqualExpected;
		this.expectThrow = expectThrow;
	}


	@Override
	public I getInput() {
		return input;
	}


	@Override
	public void setInput(I input) {
		this.input = input;
	}


	@Override
	public R getExpected() {
		return expected;
	}


	@Override
	public void setExpected(R expected) {
		this.expected = expected;
	}


	@Override
	public R getResult() {
		return result;
	}


	@Override
	public void setResult(R result) {
		this.result = result;
	}


	@Override
	public boolean isShouldInputEqualExpected() {
		return shouldInputEqualExpected;
	}


	@Override
	public void setShouldInputEqualExpected(boolean shouldInputEqualExpected) {
		this.shouldInputEqualExpected = shouldInputEqualExpected;
	}


	@Override
	public boolean isExpectThrow() {
		return expectThrow;
	}


	@Override
	public void setExpectThrow(boolean expectThrow) {
		this.expectThrow = expectThrow;
	}


	public static final <I, R> TestData<I, R> matchTrue(I input, R expected) {
		return new TestDataImpl<>(input, expected, null, true, false);
	}


	public static final <I, R> TestData<I, R> matchFalse(I input, R expected) {
		return new TestDataImpl<>(input, expected, null, false, false);
	}


	public static final <I, R> TestData<I, R> shouldThrow(I input, R expected) {
		return new TestDataImpl<>(input, expected, null, false, true);
	}

}


src/twg2/junitassist/checks/ObjectEquality.java
package twg2.junitassist.checks;

/** For testing whether a class' {@link Object#hashCode()} and {@link Object#equals(Object)}
 * methods are correctly implemented (i.e. whether the methods are reflexive, symmetric, and transitive)
 * given a group of instances of the class.
 * @author TeamworkGuy2
 * @since 2014-3-29
 */
public class ObjectEquality {
	/**
	 * @author TeamworkGuy2
	 * @since 2014-3-29
	 */
	enum EqualityTuple {
		NON_EQUAL,
		A_B_EQUAL,
		A_C_EQUAL,
		B_C_EQUAL,
		A_B_C_EQUAL,
	};


	/** Test the specified objects' {@link Object#equals(Object)}, and {@link Object#hashCode()} methods for correctness.
	 * @param a1 this object should be equal to {@code a2} and {@code a3}, but not the same object
	 * @param a2
	 * @param a3
	 * @param b1 this object should be equal to {@code b2}, but not the same object
	 * @param b2
	 * @param c1 this object should not have an equal objects among the other parameters
	 */
	public static void testEquality(Object a1, Object a2, Object a3, Object b1, Object b2, Object c1) {
		testEquality(EqualityTuple.A_B_EQUAL, b1, b2, a1);
		testEquality(EqualityTuple.A_C_EQUAL, a3, b2, a2);
		testEquality(EqualityTuple.B_C_EQUAL, a1, b2, b1);
		testEquality(EqualityTuple.A_B_C_EQUAL, a1, a2, a3);
		testEquality(EqualityTuple.NON_EQUAL, a3, b2, c1);
	}


	public static void testEquality(EqualityTuple equal, Object a, Object b, Object c) {
		a.toString();
		b.toString();
		c.toString();
		int aH = a.hashCode();
		int bH = b.hashCode();
		int cH = c.hashCode();

		switch(equal) {
		case A_B_C_EQUAL:
			// hashCode() contract
			assert aH == bH && bH == cH && aH == cH;
			// Symmetric
			assert b.equals(a) && c.equals(a) && c.equals(b);
			// Transitive
			assert a.equals(b) && b.equals(c) && a.equals(c);
			break;
		case A_B_EQUAL:
			// hashCode() contract
			assert aH == bH;
			// Symmetric
			assert b.equals(a) && !c.equals(a) && !c.equals(b);
			// Transitive
			assert a.equals(b) && !b.equals(c) && !a.equals(c);
			break;
		case A_C_EQUAL:
			// hashCode() contract
			assert aH == cH;
			// Symmetric
			assert !b.equals(a) && c.equals(a) && !c.equals(b);
			// Transitive
			assert !a.equals(b) && !b.equals(c) && a.equals(c);
			break;
		case B_C_EQUAL:
			// hashCode() contract
			assert bH == cH;
			// Symmetric
			assert !b.equals(a) && !c.equals(a) && c.equals(b);
			// Transitive
			assert !a.equals(b) && b.equals(c) && !a.equals(c);
			break;
		case NON_EQUAL:
			// hashCode() contract
			assert true;
			// Symmetric
			assert !b.equals(a) && !c.equals(a) && !c.equals(b);
			// Transitive
			assert !a.equals(b) && !b.equals(c) && !a.equals(c);
			break;
		default:
			throw new IllegalArgumentException("Unknown equality tuple type: " + equal);
		}

		// Reflexive
		assert a.equals(a) && b.equals(b) && c.equals(c);
		// "non-nullity" (Effective Java, 2nd Ed., Item 8)
		assert !a.equals(null) && !b.equals(null) && !c.equals(null);

	}

}


src/twg2/junitassist/checks/CheckCollections.java
package twg2.junitassist.checks;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;

import org.junit.Assert;

/**
 * @author TeamworkGuy2
 * @since 2015-7-12
 */
public class CheckCollections {


	public static <E> void assertLooseEquals(Iterable<? extends E> e1, Iterable<? extends E> e2) {
		assertLooseEquals(null, e1, e2);
	}


	public static <E> void assertLooseEquals(String message, Iterable<? extends E> e1, Iterable<? extends E> e2) {
		List<E> c1 = toList(e1);
		@SuppressWarnings("unchecked")
		Collection<E> c1Copy = e1 instanceof Collection ? (Collection<E>) e1 : new ArrayList<>(c1);

		List<E> c2 = toList(e2);
		@SuppressWarnings("unchecked")
		Collection<E> c2Copy = e2 instanceof Collection ? (Collection<E>) e2 : new ArrayList<>(c2);

		c1.removeAll(c2Copy);
		c2.removeAll(c1Copy);

		String msg = (message != null ? message + ": " : "") + toList(e1) + ", not equal to: " + toList(e2) + ", remaining arg1=" + c1 + ", arg2=" + c2;
		Assert.assertTrue(msg, c1.size() == 0 && c2.size() == 0);
	}


	public static <E> void assertLooseEqualsNested(Iterable<? extends Iterable<? extends E>> e1, Iterable<? extends Iterable<? extends E>> e2) {
		Iterator<? extends Iterable<? extends E>> i1 = e1.iterator();
		Iterator<? extends Iterable<? extends E>> i2 = e2.iterator();
		int baseCount = 0;
		while(i1.hasNext() && i2.hasNext()) {
			Iterable<? extends E> list1 = i1.next();
			Iterable<? extends E> list2 = i2.next();

			assertLooseEquals("sub-collection " + baseCount, list1, list2);

			baseCount++;
		}

		int count1 = baseCount;
		int count2 = baseCount;

		while(i1.hasNext()) {
			i1.next();
			count1++;
		}

		while(i2.hasNext()) {
			i2.next();
			count2++;
		}

		if(i1.hasNext() || i2.hasNext()) {
			Assert.assertTrue("collection 1 size=" + count1 + ", collection 2 size=" + count2 + " do not match", false);
		}
	}


	static final <E> List<E> toList(Iterable<? extends E> iter) {
		List<E> res = new ArrayList<>();
		for(E e : iter) {
			res.add(e);
		}
		return res;
	}

}


src/twg2/junitassist/checks/CheckArrays.java
package twg2.junitassist.checks;

import java.lang.reflect.Field;

import org.junit.Assert;

/**
 * @author TeamworkGuy2
 * @since 2016-09-18
 */
public final class CheckArrays {

	/** Check if all of the objects are non-null arrays (via {@link Class#isArray()})
	 * @param arys the arrays to check
	 * @return true if all the objects are arrays, false otherwise
	 */
	@SafeVarargs
	public static final boolean isArrays(Object... arys) {
		for(Object ary : arys) {
			if(ary == null || !ary.getClass().isArray()) {
				return false;
			}
		}
		return true;
	}


	@SafeVarargs
	public static final void assertArrayLengths(boolean[]... arys) {
		assertArrayLengths((String)null, (Object[])arys);
	}


	@SafeVarargs
	public static final void assertArrayLengths(String errMsg, boolean[]... arys) {
		assertArrayLengths(errMsg, (Object[])arys);
	}


	@SafeVarargs
	public static final void assertArrayLengths(byte[]... arys) {
		assertArrayLengths((String)null, (Object[])arys);
	}


	@SafeVarargs
	public static final void assertArrayLengths(String errMsg, byte[]... arys) {
		assertArrayLengths(errMsg, (Object[])arys);
	}


	@SafeVarargs
	public static final void assertArrayLengths(char[]... arys) {
		assertArrayLengths((String)null, (Object[])arys);
	}


	@SafeVarargs
	public static final void assertArrayLengths(String errMsg, char[]... arys) {
		assertArrayLengths(errMsg, (Object[])arys);
	}



	@SafeVarargs
	public static final void assertArrayLengths(short[]... arys) {
		assertArrayLengths((String)null, (Object[])arys);
	}


	@SafeVarargs
	public static final void assertArrayLengths(String errMsg, short[]... arys) {
		assertArrayLengths(errMsg, (Object[])arys);
	}


	@SafeVarargs
	public static final void assertArrayLengths(int[]... arys) {
		assertArrayLengths((String)null, (Object[])arys);
	}


	@SafeVarargs
	public static final void assertArrayLengths(String errMsg, int[]... arys) {
		assertArrayLengths(errMsg, (Object[])arys);
	}



	@SafeVarargs
	public static final void assertArrayLengths(long[]... arys) {
		assertArrayLengths((String)null, (Object[])arys);
	}


	@SafeVarargs
	public static final void assertArrayLengths(String errMsg, long[]... arys) {
		assertArrayLengths(errMsg, (Object[])arys);
	}


	@SafeVarargs
	public static final void assertArrayLengths(float[]... arys) {
		assertArrayLengths((String)null, (Object[])arys);
	}


	@SafeVarargs
	public static final void assertArrayLengths(String errMsg, float[]... arys) {
		assertArrayLengths(errMsg, (Object[])arys);
	}


	@SafeVarargs
	public static final void assertArrayLengths(double[]... arys) {
		assertArrayLengths((String)null, (Object[])arys);
	}


	@SafeVarargs
	public static final void assertArrayLengths(String errMsg, double[]... arys) {
		assertArrayLengths(errMsg, (Object[])arys);
	}


	@SafeVarargs
	public static final <T> void assertArrayLengths(T[]... arys) {
		assertArrayLengths(null, arys);
	}


	@SafeVarargs
	public static final <T> void assertArrayLengths(String errMsg, T[]... arys) {
		if(arys == null || arys.length > 1) {
			return;
		}
		Field aryLengthField;
		try {
			aryLengthField = arys[0].getClass().getField("length");
			if(!arys[0].getClass().isArray()) {
				throw new IllegalArgumentException("array 0 was not an array: " + arys[0]);
			}
			int firstLen = aryLengthField.getInt(arys[0]);
			for(int i = 1, size = arys.length; i < size; i++) {
				if(!arys[i].getClass().isArray()) {
					throw new IllegalArgumentException("array " + i + " was not an array: " + arys[i]);
				}
				int lenI = aryLengthField.getInt(arys[i]);
				if(lenI != firstLen) {
					Assert.assertTrue(errMsg != null ? errMsg : "array " + i + " length (" + lenI + ") not equal expected length (" + firstLen + ")", false);
				}
			}
		} catch (NoSuchFieldException e) {
			throw new RuntimeException(e);
		} catch (SecurityException e2) {
			throw new RuntimeException(e2);
		} catch (IllegalArgumentException e3) {
			throw new RuntimeException(e3);
		} catch (IllegalAccessException e4) {
			throw new RuntimeException(e4);
		}
	}


	public static final <T> void assertEquals(String errorMsg, T expect, T actual, float floatEpsilon, double doubleEpsilon) {
		Class<?> resComponent = actual.getClass().getComponentType();
		if(resComponent == Boolean.TYPE) { Assert.assertArrayEquals(errorMsg, toBooleans((boolean[])expect), toBooleans((boolean[])actual)); }
		else if(resComponent == Character.TYPE) { Assert.assertArrayEquals(errorMsg, (char[])expect, (char[])actual); }
		else if(resComponent == Byte.TYPE) { Assert.assertArrayEquals(errorMsg, (byte[])expect, (byte[])actual); }
		else if(resComponent == Short.TYPE) { Assert.assertArrayEquals(errorMsg, (short[])expect, (short[])actual); }
		else if(resComponent == Integer.TYPE) { Assert.assertArrayEquals(errorMsg, (int[])expect, (int[])actual); }
		else if(resComponent == Long.TYPE) { Assert.assertArrayEquals(errorMsg, (long[])expect, (long[])actual); }
		else if(resComponent == Float.TYPE) { Assert.assertArrayEquals(errorMsg, (float[])expect, (float[])actual, floatEpsilon); }
		else if(resComponent == Double.TYPE) { Assert.assertArrayEquals(errorMsg, (double[])expect, (double[])actual, doubleEpsilon); }
		else { Assert.assertArrayEquals(errorMsg, (Object[])expect, (Object[])actual); }
	}


	private static final Boolean[] toBooleans(boolean[] ary) {
		return toBooleans(ary, 0, ary.length);
	}


	private static final Boolean[] toBooleans(boolean[] ary, int off, int len) {
		Boolean[] res = new Boolean[len];
		for(int i = 0; i < len; i++) {
			res[i] = ary[off + i];
		}
		return res;
	}

}
